buildbot_nix/__init__.py

```diff
@@ -784,16 +784,55 @@ def __init__(self, project: GitProject, path: Path, **kwargs: Any) -> None:
         self.project = project
         self.path = path
 
+    def join_traversalsafe(self, root: Path, joined: Path) -> Path:
+        root = root.resolve()
+
+        for part in joined.parts:
+            new_root = (root / part).resolve()
+
+            if not new_root.is_relative_to(root):
+                msg = f"Joined path attempted to traverse upwards when processing {root} against {part} (gave {new_root})"
+                raise ValueError(msg)
+
+            root = new_root
+
+        return root
+
+    def join_all_traversalsafe(self, root: Path, *paths: Path) -> Path:
+        for path in paths:
+            root = self.join_traversalsafe(root, path)
+
+        return root
+
+    @defer.inlineCallbacks
     def run(self) -> Generator[Any, object, Any]:
         props = self.build.getProperties()
+
         if props.getProperty("branch") != self.project.default_branch:
             return util.SKIPPED
 
-        attr = Path(props.getProperty("attr")).name
         out_path = props.getProperty("out_path")
-        # XXX don't hardcode this
-        self.path.mkdir(parents=True, exist_ok=True)
-        (self.path / attr).write_text(out_path)
+
+        if not out_path:  # if, e.g., the build fails and doesn't produce an output
+            return util.SKIPPED
+
+        project_name = Path(props.getProperty("projectname"))
+
+        target = Path(props.getProperty("branch"))
+
+        attr = Path(props.getProperty("attr"))
```

Comment:
Should this stay the same as it was a few lines above?
```
attr = Path(props.getProperty("attr")).name
``

Reply:
I have a look.

Reply:
I thought about it a bit, and normalizing attributes, could lead to collisions, no?

Reply:
i.e. if someone does something like this:

```
{
  "foo" = ...;
  "foo/bar" = ...;
}

Reply:
Mhm. Also thinkinng about it this specific case would create collisions anyway.